libsolidity/experimental/ast/TypeSystemHelper.cpp

```diff
@@ -287,9 +288,40 @@ std::vector<experimental::Type> TypeEnvironmentHelpers::typeVars(Type _type) con
 	};
 	typeVarsImpl(_type, typeVarsImpl);
 	return typeVars;
+}
+
 
+bool TypeEnvironmentHelpers::hasGenericTypeVars(Type const& _type) const
+{
+	return ranges::any_of(
+		TypeEnvironmentHelpers{*this}.typeVars(_type),
+		[&](Type const& _maybeTypeVar) { return env.isGenericTypeVar(_maybeTypeVar); }
+	);
 }
 
+experimental::Type TypeEnvironmentHelpers::substitute(
+	Type const& _type,
+	Type const& _partToReplace,
+	Type const& _replacement
+) const
+{
+	using ranges::views::transform;
+	using ranges::to;
+
+	auto recurse = [&](experimental::Type const& _t) { return substitute(_t, _partToReplace, _replacement); };
```

Comment:
```suggestion
	auto recurse = [&](Type const& _t) { return substitute(_t, _partToReplace, _replacement); };
```
also in the other cases below

Reply:
Done.